helix-core/src/main/java/org/apache/helix/manager/zk/ParticipantManager.java

```diff
@@ -368,21 +393,35 @@ private void carryOverPreviousCurrentState() {
           continue;
         }
 
-        StateModelDefinition stateModel =
-            _dataAccessor.getProperty(_keyBuilder.stateModelDef(stateModelDefRef));
+        StateModelDefinition stateModelDef =
+            dataAccessor.getProperty(keyBuilder.stateModelDef(stateModelDefRef));
+        String initState = stateModelDef.getInitialState();
+        Map<String, String> partitionExpectedStateMap = new HashMap<>();
+        if (setToInitState) {
+          lastCurState.getPartitionStateMap().keySet()
+              .forEach(partition -> partitionExpectedStateMap.put(partition, initState));
+        } else {
+          String factoryName = lastCurState.getStateModelFactoryName();
+          StateModelFactory<? extends StateModel> stateModelFactory =
+              stateMachineEngine.getStateModelFactory(stateModelDefRef, factoryName);
+          lastCurState.getPartitionStateMap().keySet().forEach(partition -> {
+            StateModel stateModel =
+                stateModelFactory.getStateModel(lastCurState.getResourceName(), partition);
+            partitionExpectedStateMap.put(partition, stateModel.getCurrentState());
```

Comment:
NIT: let's simplify the code there. stateMode is not used in following code. Then it could be:

partitionExpectedStateMap.put(partition, stateModelFactory.getStateModel(lastCurState.getResourceName(), partition).getCurrentState());